File: packages/php/src/EnvironmentDetector.php
<?php

declare(strict_types=1);

namespace Uncaught;

/**
 * Runtime / platform environment detector for PHP.
 */
final class EnvironmentDetector
{
    private static ?array $cached = null;

    /**
     * Detect the current runtime environment.
     * Result is cached after the first invocation.
     *
     * @return array<string, string|null>
     */
    public static function detect(): array
    {
        if (self::$cached !== null) {
            return self::$cached;
        }

        $info = [];

        try {
            // Runtime
            $info['runtime'] = 'php';
            $info['runtimeVersion'] = PHP_VERSION;
            $info['platform'] = PHP_OS_FAMILY;
            $info['os'] = self::detectOS();

            // Framework detection
            self::detectFramework($info);

            // Hosting platform markers
            if (getenv('VERCEL')) {
                $info['platform'] = 'vercel';
            } elseif (getenv('RAILWAY_PROJECT_ID')) {
                $info['platform'] = 'railway';
            } elseif (getenv('FLY_APP_NAME')) {
                $info['platform'] = 'fly';
            } elseif (getenv('AWS_LAMBDA_FUNCTION_NAME')) {
                $info['platform'] = 'aws-lambda';
            } elseif (getenv('GOOGLE_CLOUD_PROJECT')) {
                $info['platform'] = 'gcp';
            }

            // Server info
            if (PHP_SAPI === 'cli') {
                $info['deviceType'] = 'server';
            }

            // Locale & timezone
            $info['locale'] = setlocale(LC_ALL, '0') ?: null;
            $info['timezone'] = date_default_timezone_get();
        } catch (\Throwable) {
            // Silent — environment detection must never throw.
        }

        self::$cached = $info;
        return $info;
    }

    /**
     * Reset the cached environment (useful for testing).
     */
    public static function resetCache(): void
    {
        self::$cached = null;
    }

    // -----------------------------------------------------------------------
    // Internal helpers
    // -----------------------------------------------------------------------

    private static function detectOS(): string
    {
        return match (PHP_OS_FAMILY) {
            'Darwin' => 'macOS',
            'Windows' => 'Windows',
            'Linux' => 'Linux',
            'BSD' => 'BSD',
            default => PHP_OS_FAMILY,
        };
    }

    private static function detectFramework(array &$info): void
    {
        // Laravel
        if (class_exists('Illuminate\Foundation\Application')) {
            $info['framework'] = 'laravel';
            try {
                $info['frameworkVersion'] = \Illuminate\Foundation\Application::VERSION;
            } catch (\Throwable) {
                // Version detection failed
            }
            return;
        }

        // Symfony
        if (class_exists('Symfony\Component\HttpKernel\Kernel')) {
            $info['framework'] = 'symfony';
            try {
                $info['frameworkVersion'] = \Symfony\Component\HttpKernel\Kernel::VERSION;
            } catch (\Throwable) {
            }
            return;
        }

        // Lumen
        if (class_exists('Laravel\Lumen\Application')) {
            $info['framework'] = 'lumen';
            return;
        }

        // Slim
        if (class_exists('Slim\App')) {
            $info['framework'] = 'slim';
            return;
        }

        // WordPress
        if (defined('ABSPATH') && defined('WPINC')) {
            $info['framework'] = 'wordpress';
            if (function_exists('get_bloginfo')) {
                $info['frameworkVersion'] = get_bloginfo('version');
            }
            return;
        }

        // Drupal
        if (defined('DRUPAL_ROOT')) {
            $info['framework'] = 'drupal';
            return;
        }
    }
}

File: packages/dotnet/src/Uncaught/EnvironmentDetector.cs
// ---------------------------------------------------------------------------
// Uncaught — runtime / platform environment detector
// ---------------------------------------------------------------------------

using System.Runtime.InteropServices;

namespace Uncaught;

/// <summary>
/// Detects the current .NET runtime environment.
/// Result is cached after the first invocation.
/// </summary>
public static class EnvironmentDetector
{
    private static EnvironmentInfo? _cached;
    private static readonly object _lock = new();

    /// <summary>
    /// Detect the current runtime environment. Result is cached.
    /// </summary>
    public static EnvironmentInfo Detect()
    {
        if (_cached != null) return _cached;

        lock (_lock)
        {
            if (_cached != null) return _cached;
            _cached = DetectInner();
            return _cached;
        }
    }

    /// <summary>
    /// Reset the cached environment (useful for testing).
    /// </summary>
    public static void ResetCache()
    {
        lock (_lock)
        {
            _cached = null;
        }
    }

    private static EnvironmentInfo DetectInner()
    {
        var info = new EnvironmentInfo();

        try
        {
            // Runtime
            info.Runtime = "dotnet";
            info.RuntimeVersion = RuntimeInformation.FrameworkDescription;
            info.Platform = RuntimeInformation.OSArchitecture.ToString().ToLower();
            info.Os = DetectOS();

            // Timezone
            try
            {
                info.Timezone = TimeZoneInfo.Local.Id;
            }
            catch
            {
                // Silent
            }

            // Locale
            try
            {
                info.Locale = System.Globalization.CultureInfo.CurrentCulture.Name;
            }
            catch
            {
                // Silent
            }

            // Framework detection
            DetectFramework(info);

            // Hosting platform markers
            if (!string.IsNullOrEmpty(System.Environment.GetEnvironmentVariable("VERCEL")))
            {
                info.Platform = "vercel";
            }
            else if (!string.IsNullOrEmpty(System.Environment.GetEnvironmentVariable("RAILWAY_PROJECT_ID")))
            {
                info.Platform = "railway";
            }
            else if (!string.IsNullOrEmpty(System.Environment.GetEnvironmentVariable("FLY_APP_NAME")))
            {
                info.Platform = "fly";
            }
            else if (!string.IsNullOrEmpty(System.Environment.GetEnvironmentVariable("AWS_LAMBDA_FUNCTION_NAME")))
            {
                info.Platform = "aws-lambda";
            }
            else if (!string.IsNullOrEmpty(System.Environment.GetEnvironmentVariable("GOOGLE_CLOUD_PROJECT")))
            {
                info.Platform = "gcp";
            }
            else if (!string.IsNullOrEmpty(System.Environment.GetEnvironmentVariable("WEBSITE_SITE_NAME")))
            {
                info.Platform = "azure";
            }
        }
        catch
        {
            // Silent — environment detection must never throw.
        }

        return info;
    }

    private static string DetectOS()
    {
        if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX)) return "macOS";
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) return "Windows";
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux)) return "Linux";
        return RuntimeInformation.OSDescription;
    }

    private static void DetectFramework(EnvironmentInfo info)
    {
        try
        {
            // Check for ASP.NET Core
            var aspnetAssembly = AppDomain.CurrentDomain
                .GetAssemblies()
                .FirstOrDefault(a => a.GetName().Name == "Microsoft.AspNetCore");

            if (aspnetAssembly != null)
            {
                info.Framework = "aspnet-core";
                info.FrameworkVersion = aspnetAssembly.GetName().Version?.ToString();
                return;
            }

            // Check for Blazor
            var blazorAssembly = AppDomain.CurrentDomain
                .GetAssemblies()
                .FirstOrDefault(a => a.GetName().Name == "Microsoft.AspNetCore.Components");

            if (blazorAssembly != null)
            {
                info.Framework = "blazor";
                info.FrameworkVersion = blazorAssembly.GetName().Version?.ToString();
                return;
            }

            // Check for MAUI
            var mauiAssembly = AppDomain.CurrentDomain
                .GetAssemblies()
                .FirstOrDefault(a => a.GetName().Name == "Microsoft.Maui");

            if (mauiAssembly != null)
            {
                info.Framework = "maui";
                info.FrameworkVersion = mauiAssembly.GetName().Version?.ToString();
                return;
            }
        }
        catch
        {
            // Silent
        }
    }
}
